### PR TITLE
[RLlib] Solve PyTorch/TF-eager A3C async race condition between calling model and its value function.

### PR DESCRIPTION
**Pull request by sven1977**
Simple solution for PyTorch and TF-eager A3C async race condition happening when:
- Calling model and its value function in succession (as done in e.g. a loss function).
- In between the two calls above, there is e.g. an action_computation.
- This usually causes a shape error b/c the value_function call now returns the value for the single-action computation instead of the loss batch.

The above case only happens if `sample_async=True` (sampling and loss calculations are done on the same worker, like in A3C).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/a3c/a3c_torch_policy.py [hunk 1]
```diff
@@ -9,8 +11,10 @@
 
 
 def actor_critic_loss(policy, model, dist_class, train_batch):
-    logits, _ = model.from_batch(train_batch)
-    values = model.value_function()
+    with policy._model_lock:
```
Comment (ericl) [inline rllib/agents/a3c/a3c_torch_policy.py:14]: Instead of inserting the model lock inside the policy loss (which is kind of violating the abstraction since the loss shouldn't know about the lock), why not add the locking to TorchPolicy itself?

I.e., add a global lock used by all methods of torch policy.

**Review by ericl — CHANGES_REQUESTED** [review]
See lock comment

File: release/long_running_tests/run.sh [hunk 2]
```diff
@@ -52,7 +52,7 @@ wheel="https://s3-us-west-2.amazonaws.com/ray-wheels/$ray_branch/$commit/ray-$ra
 
 # Serve load testing tool
 cur_dir=$(pwd)
-cd /tmp && rm -rf wrk && git clone https://github.com/wg/wrk.git wrk && cd wrk && make -j && cp wrk /usr/local/bin
+cd /tmp && rm -rf wrk && git clone https://github.com/wg/wrk.git wrk && cd wrk && make -j && sudo cp wrk /usr/local/bin
```
Comment (ericl) [inline release/long_running_tests/run.sh:55]: Is this related to this pr?

File: rllib/policy/torch_policy.py [hunk 3]
```diff
@@ -259,6 +264,8 @@ def _compute_action_helper(self, input_dict, state_batches, seq_lens,
         extra_fetches = self.extra_action_out(input_dict, state_batches,
                                               self.model, action_dist)
 
+        self._lock.release()
```
Comment (ericl) [inline rllib/policy/torch_policy.py:267]: Hmm can we do the release in a finally block to avoid hangs? This might cause a lot of indentation changes, so one strategy is to use an method annotation (e.g.,

```
@with_lock
def _compute_action_helper:...

```

**Comment by sven1977** [comment]
tf-eager has the same problem, btw. I'll add it to this PR as well.

File: release/long_running_tests/run.sh (hunk 2 above)
Comment (sven1977) [reply release/long_running_tests/run.sh:55]: done

**Comment by sven1977** [comment]
Tests are hanging ...

**Comment by sven1977** [comment]
@ericl This is ready for review again. All tests are passing now.

**Review by ericl — APPROVED** [review]
